File: GithubUserApp2/app/src/main/java/com/example/githubuserapp2/DetailActivity.java
package com.example.githubuserapp2;

import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;
import androidx.viewpager2.widget.ViewPager2;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.util.Log;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;
import com.loopj.android.http.AsyncHttpClient;
import com.loopj.android.http.AsyncHttpResponseHandler;

import org.json.JSONObject;

import cz.msebera.android.httpclient.Header;

public class DetailActivity extends AppCompatActivity {
    public static final String EXTRA_USER = "extra_user_data";
    String name, user_company, repo, location;
    TextView tvName, tvUsername, tvLocation, tvRepo, tvCompany;
    ImageView ivAvatar;

    @StringRes
    private final int[] TAB_TITLES = new int[]{
            R.string.tab_text_1,
            R.string.tab_text_2
    };

    private static final String TAG = DetailActivity.class.getSimpleName();


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_detail);

        getSupportActionBar().setTitle("Detail User");

        User user = getIntent().getParcelableExtra(EXTRA_USER);

        ivAvatar = findViewById(R.id.img_detail_avatar);
        tvUsername = findViewById(R.id.tv_detail_username);
        tvName = findViewById(R.id.tv_detail_name);
        tvLocation = findViewById(R.id.tv_detail_location);
        tvRepo = findViewById(R.id.tv_detail_repository);
        tvCompany = findViewById(R.id.tv_detail_company);

        final ProgressDialog progress = new ProgressDialog(DetailActivity.this);
        progress.setMessage(getString(R.string.progress));
        progress.show();

        if (user != null) {
            Glide.with(DetailActivity.this)
                    .load(user.getPhoto())
                    .into(ivAvatar);

            getDetailUser(user.getUsername(), progress);
            tabLayout();
        }
    }

    private void getDetailUser(final String username, ProgressDialog progress) {
        AsyncHttpClient client = new AsyncHttpClient();

        String url = "https://api.github.com/users/" + username;

        client.addHeader("Authorization", "token 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");
        client.addHeader("User-Agent", "request");
        client.get(url, new AsyncHttpResponseHandler() {
            @Override
            public void onSuccess(int statusCode, Header[] headers, byte[] responseBody) {
                String result = new String(responseBody);
                Log.d(TAG, result);
                try {
                    JSONObject jsonObject = new JSONObject(result);
                    name = jsonObject.getString("name");
                    user_company = jsonObject.getString("company");
                    location = jsonObject.getString("location");
                    repo = jsonObject.getString("public_repos");

                    if (user_company == "null"){
                        user_company = " ";
                    }
                    if (location == "null"){
                        location = " ";
                    }
                    if (name == "null"){
                        name = " ";
                    }

                    tvUsername.setText(username);
                    tvName.setText(name);
                    tvLocation.setText(getString(R.string.txt_location) + location);
                    tvRepo.setText(getString(R.string.txt_repository) + repo);
                    tvCompany.setText(getString(R.string.txt_company) + user_company);
                    progress.dismiss();

                } catch (Exception e) {
                    Toast.makeText(DetailActivity.this, e.getMessage(), Toast.LENGTH_SHORT).show();
                    e.printStackTrace();
                }

            }

            @Override
            public void onFailure(int statusCode, Header[] headers, byte[] responseBody, Throwable error) {
                progress.dismiss();
                String errorMessage;
                switch (statusCode) {
                    case 401:
                        errorMessage = statusCode + " : Bad Request";
                        break;
                    case 403:
                        errorMessage = statusCode + " : Forbiden";
                        break;
                    case 404:
                        errorMessage = statusCode + " : Not Found";
                        break;
                    default:
                        errorMessage =  statusCode + " : " + error.getMessage();
                        break;
                }
                Toast.makeText(DetailActivity.this, errorMessage, Toast.LENGTH_SHORT).show();
            }
        });
    }


    private void tabLayout(){
        SectionsPagerAdapter sectionsPagerAdapter = new SectionsPagerAdapter(this);
        ViewPager2 viewPager = findViewById(R.id.view_pager);
        viewPager.setAdapter(sectionsPagerAdapter);
        TabLayout tabs = findViewById(R.id.tabs);
        new TabLayoutMediator(tabs, viewPager,
                (tab, position) -> tab.setText(getResources().getString(TAB_TITLES[position]))
        ).attach();
        if(getSupportActionBar() != null) {
            getSupportActionBar().setElevation(0);
        }
    }
}

File: GithubUserApps3/app/src/main/java/com/example/githubuserapps3/DetailActivity.kt
package com.example.githubuserapps3

import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.AppCompatImageButton
import androidx.lifecycle.lifecycleScope
import androidx.viewpager.widget.ViewPager
import com.bumptech.glide.Glide
import com.example.githubuserapps3.adapter.FavoriteAdapter
import com.example.githubuserapps3.adapter.SectionsPagerAdapter
import com.example.githubuserapps3.data.Favorite
import com.example.githubuserapps3.data.User
import com.example.githubuserapps3.db.DatabaseContract
import com.example.githubuserapps3.db.FavoriteHelper
import com.example.githubuserapps3.helper.MappingHelper
import com.google.android.material.tabs.TabLayout
import com.loopj.android.http.AsyncHttpClient
import com.loopj.android.http.AsyncHttpResponseHandler
import cz.msebera.android.httpclient.Header
import de.hdodenhof.circleimageview.CircleImageView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import org.json.JSONObject


lateinit var dcontext: Context
class DetailActivity : AppCompatActivity(), View.OnClickListener {

    companion object {
        private val TAG = DetailActivity::class.java.simpleName
        const val EXTRA_NAME = "extra_name"
        const val EXTRA_POSITION = "extra_position"
    }

    private lateinit var ivAvatar: CircleImageView
    private lateinit var tvUsername: TextView
    private lateinit var tvName: TextView
    private lateinit var tvLocation: TextView
    private lateinit var tvRepo: TextView
    private lateinit var tvCompany: TextView
    private lateinit var viewPager: ViewPager
    private lateinit var tabs: TabLayout
    private lateinit var buttonFavorite: AppCompatImageButton

    private var position: Int? = null
    private lateinit var favoriteHelper: FavoriteHelper
    private var isFavorite: Boolean = false

    private lateinit var username: String
    private lateinit var avatar: String


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail)
        dcontext = this

        ivAvatar = findViewById(R.id.detail_avatar);
        tvUsername = findViewById(R.id.detail_username);
        tvName = findViewById(R.id.detail_name);
        tvLocation = findViewById(R.id.detail_location);
        tvRepo = findViewById(R.id.detail_repository);
        tvCompany = findViewById(R.id.detail_company);
        viewPager = findViewById(R.id.view_pager)
        tabs = findViewById(R.id.tabs)
        buttonFavorite = findViewById(R.id.btn_favorite)

        username = intent.getStringExtra(EXTRA_NAME).toString()
        isFavorite = checkFavorite()

        if (isFavorite) {
            position = intent.getIntExtra(EXTRA_POSITION, -1)
            val checked: Int = R.drawable.ic_baseline_favorite_red_24
            buttonFavorite.setImageResource(checked)
        } else {
            val unChecked: Int = R.drawable.ic_baseline_favorite_border_24
            buttonFavorite.setImageResource(unChecked)
        }


        buttonFavorite.setOnClickListener(this)
        supportActionBar?.title = username
        getDetailUser(username)
        viewPagerConfig()
    }

    private fun checkFavorite(): Boolean {
        var isTrue = false
        val favoriteHelper = FavoriteHelper.getInstance(applicationContext)
        favoriteHelper.open()
        val cursor = favoriteHelper.queryByName(username)
        val favorite =  MappingHelper.mapCursorToArrayList(cursor)
        if (favorite.size > 0) {
            isTrue = true
        }
        favoriteHelper.close()
        return isTrue
    }


    override fun onClick(view: View) {
        val checked: Int = R.drawable.ic_baseline_favorite_red_24
        val unChecked: Int = R.drawable.ic_baseline_favorite_border_24
        if (view.id == R.id.btn_favorite) {
            favoriteHelper = FavoriteHelper.getInstance(applicationContext)
            favoriteHelper.open()

            if (isFavorite) {
                buttonFavorite.setImageResource(unChecked)
                isFavorite = false

                val result = favoriteHelper.deleteById(position.toString()).toLong()
                if (result > 0) {
                    Toast.makeText(this@DetailActivity, "Berhasil menghapus dari favorit", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(this@DetailActivity, "Gagal menghapus", Toast.LENGTH_SHORT).show()
                }
            }
            else {
                buttonFavorite.setImageResource(checked)
                isFavorite = true

                val values = ContentValues()
                values.put(DatabaseContract.UserColumns.USERNAME, username)
                values.put(DatabaseContract.UserColumns.AVATAR, avatar)

                val result = favoriteHelper.insert(values)


                if (result > 0) {
                    position = result.toInt()
                    Toast.makeText(this@DetailActivity, "Berhasil ditambahkan ke favorit", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(this@DetailActivity, "Gagal menambah data", Toast.LENGTH_SHORT).show()
                }
            }

            favoriteHelper.close()
        }
    }

    private fun viewPagerConfig() {
        val sectionsPagerAdapter = SectionsPagerAdapter(this, supportFragmentManager)
        viewPager.adapter = sectionsPagerAdapter
        tabs.setupWithViewPager(viewPager)

        supportActionBar?.elevation = 0f
    }


    private fun getDetailUser(username: String) {
        val client = AsyncHttpClient()
        val clientSecret = "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"
        val url =
            "https://api.github.com/users/$username"

        client.addHeader("Authorization", "token $clientSecret")
        client.addHeader("User-Agent", "request")
        client.get(url, object : AsyncHttpResponseHandler() {
            @SuppressLint("SetTextI18n")
            override fun onSuccess(
                statusCode: Int,
                headers: Array<Header>,
                responseBody: ByteArray
            ) {
                val result = String(responseBody)
                Log.d(DetailActivity.TAG, result)
                try {
                    val jsonObject = JSONObject(result)
                    avatar = jsonObject.getString("avatar_url")
                    var name = jsonObject.getString("name")
                    var userCompany = jsonObject.getString("company")
                    var location = jsonObject.getString("location")
                    val repo = jsonObject.getString("public_repos")

                    if (userCompany == "null") {
                        userCompany = " ";
                    }
                    if (location == "null") {
                        location = " ";
                    }
                    if (name == "null") {
                        name = " ";
                    }
                    Log.d("data", avatar)

                    Glide.with(dcontext).load(avatar).into(ivAvatar)
                    tvUsername.setText(username)
                    tvName.setText(name)
                    tvCompany.text = getString(R.string.company) + " " + userCompany
                    tvLocation.text = getString(R.string.location) + " " + location
                    tvRepo.text = getString(R.string.repository) + " " + repo

                } catch (e: Exception) {
                    Toast.makeText(this@DetailActivity, e.message, Toast.LENGTH_SHORT).show()
                    e.printStackTrace()
                }

            }

            override fun onFailure(
                statusCode: Int,
                headers: Array<Header>,
                responseBody: ByteArray,
                error: Throwable
            ) {
                val errorMessage = when (statusCode) {
                    401 -> "$statusCode : Bad Request"
                    403 -> "$statusCode : Forbidden"
                    404 -> "$statusCode : Not Found"
                    else -> "$statusCode : ${error.message}"
                }
                Toast.makeText(this@DetailActivity, errorMessage, Toast.LENGTH_SHORT).show()
            }
        })
    }
}